docs/CODE_GUIDELINES.md

```diff
@@ -198,80 +217,196 @@ exampleClass      ->InitExample();
 biggerExampleClass->InitExample();
 ```
 
-Right:
+**[back to top](#table-of-contents)**
+
+## 4. Statements
+
+Do not put multiple statements on a single line. Always use a new line for a new statement. It is much easier to debug if one can pinpoint a precise line number.
+
+✅ Good:
+```cpp
+std::vector<std::string> test;
+test.push_back("foobar");
+```
+
+❌ Bad:
 ```cpp
-int value1 = 0;
-int value2 = 0;
-CExampleClass *exampleClass = nullptr;
-CBiggerExampleClass *biggerExampleClass = nullptr;
-exampleClass = new CExampleClass(value1, value2);
-biggerExampleClass = new CBiggerExampleClass(value1, value2);
-exampleClass->InitExample();
-biggerExampleClass->InitExample();
+std::vector<std::string> test; test.push_back("foobar");
 ```
 
 **[back to top](#table-of-contents)**
 
-## 10. Controls Statements
-Insert a new line before every:
-* else in an if statement
-* catch in a try statement
-* while in a do statement
+## 5. Declarations
 
-### 10.1. if else
-Put `then`, `return` or `throw` statements on a new line. Keep `else if` statements on one line.
+### 5.1. Multiple declarations
+Do not put multiple declarations on a single line. This avoids confusion with differing pointers, references, and initialization values on the same line (cf. [ISO C++ guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es10-declare-one-name-only-per-declaration)).
+
+✅ Good:
 ```cpp
-if (true)
-  return;
-if (true)
-{
-  [...]
-} 
-else if (false)
+char* a;
+char b;
+```
+
+❌ Bad:
+```cpp
+char* a, b;
+```
+
+### 5.2. Pointer and reference types
+Left-align `*` and `&` to the base type they modify.
+
+✅ Good:
+```cpp
+char* a;
+void test(const std::string& b);
+```
+
+❌ Bad:
+```cpp
+char *a;
+char * b;
+void test(const std::string &b);
+```
+
+### 5.3. `const` and other modifiers
+Place `const` and similar modifiers in front of the base type.
+
+✅ Good:
+```cpp
+void test(const std::string& a);
+```
+
+❌ Bad:
+```cpp
+void test(std::string const& a);
```

Comment:
This brings up two points - we want to avoid const on primitive parameters, e.g. `test(const int a)`, and also we allow const after the modifier when we actually want const values, e.g.

```cpp
class Foo
{
  const Bar* const m_immutable;
};
```

Reply:
> This brings up two points - we want to avoid const on primitive parameters, e.g. test(const int a)

Don't have a strong opinion on this one - I think it's fine to do so if you want to emphasize that the function should not change the variable in its body (even if it would not be visible to the caller anyway).

> and also we allow const after the modifier when we actually want const values

Yeah that's a different type of `const` but maybe I can improve the wording.

Reply:
> Don't have a strong opinion on this one - I think it's fine to do so if you want to emphasize that the function should not change the variable in its body (even if it would not be visible to the caller anyway).

I am against this, because I don't think this internal detail should be exposed to external callers. However, inside a function, I encourage the use of const variables as much as possible -- see how rust defaults everything to const and requires `mutable` to be declared explicitly.